diff --git a/flux.hpp b/flux.hpp
index 20ff410..a1b4c4c 100644
--- a/flux.hpp
+++ b/flux.hpp
@@ -1014,7 +1014,7 @@ namespace Flux {
 
         static void load_from_file_and_test(const std::string& file_path) {
             // ggml_backend_t backend    = ggml_backend_cuda_init(0);
-            ggml_backend_t backend           = ggml_backend_cpu_init();
+            ggml_backend_t backend           = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
             ggml_type model_data_type        = GGML_TYPE_Q8_0;
             std::shared_ptr<FluxRunner> flux = std::shared_ptr<FluxRunner>(new FluxRunner(backend));
             {
diff --git a/ggml_extend.hpp b/ggml_extend.hpp
index 236e7c1..6a5a5c7 100644
--- a/ggml_extend.hpp
+++ b/ggml_extend.hpp
@@ -1001,16 +1001,8 @@ __STATIC_INLINE__ struct ggml_tensor* ggml_nn_group_norm(struct ggml_context* ct
 }
 
 __STATIC_INLINE__ void ggml_backend_tensor_get_and_sync(ggml_backend_t backend, const struct ggml_tensor* tensor, void* data, size_t offset, size_t size) {
-#if defined(SD_USE_CUDA) || defined(SD_USE_SYCL)
-    if (!ggml_backend_is_cpu(backend)) {
-        ggml_backend_tensor_get_async(backend, tensor, data, offset, size);
-        ggml_backend_synchronize(backend);
-    } else {
-        ggml_backend_tensor_get(tensor, data, offset, size);
-    }
-#else
-    ggml_backend_tensor_get(tensor, data, offset, size);
-#endif
+    ggml_backend_tensor_get_async(backend, tensor, data, offset, size);
+    ggml_backend_synchronize(backend);
 }
 
 __STATIC_INLINE__ float ggml_backend_tensor_get_f32(ggml_tensor* tensor) {
@@ -1140,6 +1132,7 @@ protected:
     std::map<struct ggml_tensor*, const void*> backend_tensor_data_map;
 
     ggml_backend_t backend = NULL;
+    enum ggml_backend_dev_type backend_dev_type = GGML_BACKEND_DEVICE_TYPE_CPU;
 
     void alloc_params_ctx() {
         struct ggml_init_params params;
@@ -1196,7 +1189,7 @@ protected:
         LOG_DEBUG("%s compute buffer size: %.2f MB(%s)",
                   get_desc().c_str(),
                   compute_buffer_size / 1024.0 / 1024.0,
-                  ggml_backend_is_cpu(backend) ? "RAM" : "VRAM");
+                  backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM");
         return true;
     }
 
@@ -1216,6 +1209,7 @@ public:
 
     GGMLRunner(ggml_backend_t backend)
         : backend(backend) {
+        backend_dev_type = ggml_backend_dev_type(ggml_backend_get_device(backend));
         alloc_params_ctx();
     }
 
@@ -1244,12 +1238,12 @@ public:
         LOG_DEBUG("%s params backend buffer size = % 6.2f MB(%s) (%i tensors)",
                   get_desc().c_str(),
                   params_buffer_size / (1024.0 * 1024.0),
-                  ggml_backend_is_cpu(backend) ? "RAM" : "VRAM",
+                  backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU  ? "RAM" : "VRAM",
                   num_tensors);
         // printf("%s params backend buffer size = % 6.2f MB(%s) (%i tensors)\n",
         //           get_desc().c_str(),
         //           params_buffer_size / (1024.0 * 1024.0),
-        //           ggml_backend_is_cpu(backend) ? "RAM" : "VRAM",
+        //           backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM",
         //           num_tensors);
         return true;
     }
@@ -1286,7 +1280,7 @@ public:
             return NULL;
         }
         // it's performing a compute, check if backend isn't cpu
-        if (!ggml_backend_is_cpu(backend) && (tensor->buffer == NULL || ggml_backend_buffer_is_host(tensor->buffer))) {
+        if (backend_dev_type != GGML_BACKEND_DEVICE_TYPE_CPU && (tensor->buffer == NULL || ggml_backend_buffer_is_host(tensor->buffer))) {
             // pass input tensors to gpu memory
             auto backend_tensor = ggml_dup_tensor(compute_ctx, tensor);
 
diff --git a/mmdit.hpp b/mmdit.hpp
index dee7b1c..50feccf 100644
--- a/mmdit.hpp
+++ b/mmdit.hpp
@@ -970,7 +970,7 @@ struct MMDiTRunner : public GGMLRunner {
 
     static void load_from_file_and_test(const std::string& file_path) {
         // ggml_backend_t backend    = ggml_backend_cuda_init(0);
-        ggml_backend_t backend             = ggml_backend_cpu_init();
+        ggml_backend_t backend             = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
         ggml_type model_data_type          = GGML_TYPE_F16;
         std::shared_ptr<MMDiTRunner> mmdit = std::shared_ptr<MMDiTRunner>(new MMDiTRunner(backend));
         {
diff --git a/model.cpp b/model.cpp
index d9cf7b4..1d69efb 100644
--- a/model.cpp
+++ b/model.cpp
@@ -2133,7 +2133,7 @@ bool ModelLoader::save_to_gguf_file(const std::string& file_path, ggml_type outt
     LOG_INFO("save t5xxl weight type: %s", ggml_type_name(t5xxl_outtype));
     LOG_INFO("save diffusion model weight type: %s", ggml_type_name(outtype));
 
-    auto backend    = ggml_backend_cpu_init();
+    auto backend    = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
     size_t mem_size = 1 * 1024 * 1024;  // for padding
     mem_size += tensor_storages.size() * ggml_tensor_overhead();
     mem_size += get_params_mem_size(backend, outtype, vae_outtype, clip_l_outtype, clip_g_outtype, t5xxl_outtype);
diff --git a/preprocessing.hpp b/preprocessing.hpp
index 4ea1dba..d190ab7 100644
--- a/preprocessing.hpp
+++ b/preprocessing.hpp
@@ -15,6 +15,7 @@ void convolve(struct ggml_tensor* input, struct ggml_tensor* output, struct ggml
     ggml_tensor* h  = ggml_conv_2d(ctx0, kernel_fp16, input, 1, 1, padding, padding, 1, 1);
     ggml_cgraph* gf = ggml_new_graph(ctx0);
     ggml_build_forward_expand(gf, ggml_cpy(ctx0, h, output));
+
     ggml_graph_compute_with_ctx(ctx0, gf, 1);
     ggml_free(ctx0);
 }
@@ -224,4 +225,4 @@ uint8_t* preprocess_canny(uint8_t* img, int width, int height, float high_thresh
     return output;
 }
 
-#endif  // __PREPROCESSING_HPP__
\ No newline at end of file
+#endif  // __PREPROCESSING_HPP__
diff --git a/stable-diffusion.cpp b/stable-diffusion.cpp
index 0ea4c2a..e07bd09 100644
--- a/stable-diffusion.cpp
+++ b/stable-diffusion.cpp
@@ -142,6 +142,7 @@ void calculate_alphas_cumprod(float* alphas_cumprod,
 
 class StableDiffusionGGML {
 public:
+    ggml_backend_t backend_cpu         = NULL;
     ggml_backend_t backend             = NULL;  // general backend
     ggml_backend_t clip_backend        = NULL;
     ggml_backend_t vae_backend         = NULL;
@@ -186,6 +187,10 @@ public:
 
     std::shared_ptr<Denoiser> denoiser = std::make_shared<CompVisDenoiser>();
 
+    enum ggml_backend_dev_type backend_dev_type = GGML_BACKEND_DEVICE_TYPE_CPU;
+    enum ggml_backend_dev_type clip_backend_dev_type = GGML_BACKEND_DEVICE_TYPE_CPU;
+    enum ggml_backend_dev_type vae_backend_dev_type = GGML_BACKEND_DEVICE_TYPE_CPU;
+    enum ggml_backend_dev_type control_net_backend_dev_type = GGML_BACKEND_DEVICE_TYPE_CPU;
     StableDiffusionGGML() = default;
 
     StableDiffusionGGML(int n_threads,
@@ -275,10 +280,12 @@ public:
             }
         }
 
+        backend_cpu = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
+
         // initialize the backend
         if (gpu_devices.empty()) {
             // no GPU devices available
-            backend = ggml_backend_cpu_init();
+            backend = backend_cpu;
         } else if (gpu_devices.size() < 3) {
             // use the last GPU device: device 0, device 1
             backend = ggml_backend_dev_init(gpu_devices[gpu_devices.size() - 1].first, nullptr);
@@ -298,17 +305,17 @@ public:
                 clip_backend = backend;
                 if (clip_on_cpu) {
                     LOG_INFO("CLIP: Using CPU backend");
-                    clip_backend = ggml_backend_cpu_init();
+                    clip_backend = backend_cpu;
                 }
                 vae_backend = backend;
                 if (vae_on_cpu) {
                     LOG_INFO("VAE Autoencoder: Using CPU backend");
-                    vae_backend = ggml_backend_cpu_init();
+                    vae_backend = backend_cpu;
                 }
                 control_net_backend = backend;
                 if (control_net_cpu) {
                     LOG_INFO("ControlNet: Using CPU backend");
-                    control_net_backend = ggml_backend_cpu_init();
+                    control_net_backend = backend_cpu;
                 }
                 break;
             }
@@ -316,19 +323,19 @@ public:
                 // device 0: clip, vae, control_net
                 if (clip_on_cpu) {
                     LOG_INFO("CLIP: Using CPU backend");
-                    clip_backend = ggml_backend_cpu_init();
+                    clip_backend = backend_cpu;
                 } else {
                     clip_backend = ggml_backend_dev_init(gpu_devices[0].first, nullptr);
                 }
                 if (vae_on_cpu) {
                     LOG_INFO("VAE Autoencoder: Using CPU backend");
-                    vae_backend = ggml_backend_cpu_init();
+                    vae_backend = backend_cpu;
                 } else {
                     vae_backend = ggml_backend_dev_init(gpu_devices[0].first, nullptr);
                 }
                 if (control_net_cpu) {
                     LOG_INFO("ControlNet: Using CPU backend");
-                    control_net_backend = ggml_backend_cpu_init();
+                    control_net_backend = backend_cpu;
                 } else {
                     control_net_backend = ggml_backend_dev_init(gpu_devices[0].first, nullptr);
                 }
@@ -339,25 +346,35 @@ public:
                 // device 1: vae
                 if (clip_on_cpu) {
                     LOG_INFO("CLIP: Using CPU backend");
-                    clip_backend = ggml_backend_cpu_init();
+                    clip_backend = backend_cpu;
                 } else {
                     clip_backend = ggml_backend_dev_init(gpu_devices[0].first, nullptr);
                 }
                 if (vae_on_cpu) {
                     LOG_INFO("VAE Autoencoder: Using CPU backend");
-                    vae_backend = ggml_backend_cpu_init();
+                    vae_backend = backend_cpu;
                 } else {
                     vae_backend = ggml_backend_dev_init(gpu_devices[1].first, nullptr);
                 }
                 if (control_net_cpu) {
                     LOG_INFO("ControlNet: Using CPU backend");
-                    control_net_backend = ggml_backend_cpu_init();
+                    control_net_backend = backend_cpu;
                 } else {
                     control_net_backend = ggml_backend_dev_init(gpu_devices[0].first, nullptr);
                 }
             }
         }
 
+        backend_dev_type = ggml_backend_dev_type(ggml_backend_get_device(backend));
+        if (clip_backend) {
+            clip_backend_dev_type = ggml_backend_dev_type(ggml_backend_get_device(clip_backend));
+        }
+        if (vae_backend) {
+            vae_backend_dev_type = ggml_backend_dev_type(ggml_backend_get_device(vae_backend));
+        }
+        if (control_net_backend) {
+            control_net_backend_dev_type = ggml_backend_dev_type(ggml_backend_get_device(control_net_backend));
+        }
         ModelLoader model_loader;
 
         vae_tiling = vae_tiling_;
@@ -669,25 +686,25 @@ public:
 
             size_t total_params_ram_size  = 0;
             size_t total_params_vram_size = 0;
-            if (ggml_backend_is_cpu(clip_backend)) {
+            if (clip_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU) {
                 total_params_ram_size += clip_params_mem_size + pmid_params_mem_size;
             } else {
                 total_params_vram_size += clip_params_mem_size + pmid_params_mem_size;
             }
 
-            if (ggml_backend_is_cpu(backend)) {
+            if (backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU) {
                 total_params_ram_size += unet_params_mem_size;
             } else {
                 total_params_vram_size += unet_params_mem_size;
             }
 
-            if (ggml_backend_is_cpu(vae_backend)) {
+            if (vae_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU) {
                 total_params_ram_size += vae_params_mem_size;
             } else {
                 total_params_vram_size += vae_params_mem_size;
             }
 
-            if (ggml_backend_is_cpu(control_net_backend)) {
+            if (control_net_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU) {
                 total_params_ram_size += control_net_params_mem_size;
             } else {
                 total_params_vram_size += control_net_params_mem_size;
@@ -701,15 +718,15 @@ public:
                 total_params_vram_size / 1024.0 / 1024.0,
                 total_params_ram_size / 1024.0 / 1024.0,
                 clip_params_mem_size / 1024.0 / 1024.0,
-                ggml_backend_is_cpu(clip_backend) ? "RAM" : "VRAM",
+                clip_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM",
                 unet_params_mem_size / 1024.0 / 1024.0,
-                ggml_backend_is_cpu(backend) ? "RAM" : "VRAM",
+                backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM",
                 vae_params_mem_size / 1024.0 / 1024.0,
-                ggml_backend_is_cpu(vae_backend) ? "RAM" : "VRAM",
+                vae_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM",
                 control_net_params_mem_size / 1024.0 / 1024.0,
-                ggml_backend_is_cpu(control_net_backend) ? "RAM" : "VRAM",
+                control_net_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM",
                 pmid_params_mem_size / 1024.0 / 1024.0,
-                ggml_backend_is_cpu(clip_backend) ? "RAM" : "VRAM");
+                clip_backend_dev_type == GGML_BACKEND_DEVICE_TYPE_CPU ? "RAM" : "VRAM");
         }
 
         int64_t t1 = ggml_time_ms();
diff --git a/t5.hpp b/t5.hpp
index 2a53e27..e5e59a5 100644
--- a/t5.hpp
+++ b/t5.hpp
@@ -954,7 +954,7 @@ struct T5Embedder {
 
     static void load_from_file_and_test(const std::string& file_path) {
         // ggml_backend_t backend    = ggml_backend_cuda_init(0);
-        ggml_backend_t backend         = ggml_backend_cpu_init();
+        ggml_backend_t backend         = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
         ggml_type model_data_type      = GGML_TYPE_F32;
         std::shared_ptr<T5Embedder> t5 = std::shared_ptr<T5Embedder>(new T5Embedder(backend));
         {
diff --git a/upscaler.cpp b/upscaler.cpp
index a808638..dfa6aba 100644
--- a/upscaler.cpp
+++ b/upscaler.cpp
@@ -57,7 +57,7 @@ struct UpscalerGGML {
         // initialize the backend
         if (gpu_devices.empty()) {
             // no GPU devices available
-            backend = ggml_backend_cpu_init();
+            backend = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
         } else {
             // use the first GPU device: device 0
             backend = ggml_backend_dev_init(gpu_devices[0].first, nullptr);
diff --git a/util.cpp b/util.cpp
index 8c358dd..687cc57 100644
--- a/util.cpp
+++ b/util.cpp
@@ -413,20 +413,20 @@ void sd_set_progress_callback(sd_progress_cb_t cb, void* data) {
 const char* sd_get_system_info() {
     static char buffer[1024];
     std::stringstream ss;
-    ss << "System Info: \n";
-    ss << "    SSE3 = " << ggml_cpu_has_sse3() << std::endl;
-    ss << "    AVX = " << ggml_cpu_has_avx() << std::endl;
-    ss << "    AVX2 = " << ggml_cpu_has_avx2() << std::endl;
-    ss << "    AVX512 = " << ggml_cpu_has_avx512() << std::endl;
-    ss << "    AVX512_VBMI = " << ggml_cpu_has_avx512_vbmi() << std::endl;
-    ss << "    AVX512_VNNI = " << ggml_cpu_has_avx512_vnni() << std::endl;
-    ss << "    FMA = " << ggml_cpu_has_fma() << std::endl;
-    ss << "    NEON = " << ggml_cpu_has_neon() << std::endl;
-    ss << "    ARM_FMA = " << ggml_cpu_has_arm_fma() << std::endl;
-    ss << "    F16C = " << ggml_cpu_has_f16c() << std::endl;
-    ss << "    FP16_VA = " << ggml_cpu_has_fp16_va() << std::endl;
-    ss << "    WASM_SIMD = " << ggml_cpu_has_wasm_simd() << std::endl;
-    ss << "    VSX = " << ggml_cpu_has_vsx() << std::endl;
+//    ss << "System Info: \n";
+//    ss << "    SSE3 = " << ggml_cpu_has_sse3() << std::endl;
+//    ss << "    AVX = " << ggml_cpu_has_avx() << std::endl;
+//    ss << "    AVX2 = " << ggml_cpu_has_avx2() << std::endl;
+//    ss << "    AVX512 = " << ggml_cpu_has_avx512() << std::endl;
+//    ss << "    AVX512_VBMI = " << ggml_cpu_has_avx512_vbmi() << std::endl;
+//    ss << "    AVX512_VNNI = " << ggml_cpu_has_avx512_vnni() << std::endl;
+//    ss << "    FMA = " << ggml_cpu_has_fma() << std::endl;
+//    ss << "    NEON = " << ggml_cpu_has_neon() << std::endl;
+//    ss << "    ARM_FMA = " << ggml_cpu_has_arm_fma() << std::endl;
+//    ss << "    F16C = " << ggml_cpu_has_f16c() << std::endl;
+//    ss << "    FP16_VA = " << ggml_cpu_has_fp16_va() << std::endl;
+//    ss << "    WASM_SIMD = " << ggml_cpu_has_wasm_simd() << std::endl;
+//    ss << "    VSX = " << ggml_cpu_has_vsx() << std::endl;
     snprintf(buffer, sizeof(buffer), "%s", ss.str().c_str());
     return buffer;
 }
